partialmessages/partialmsgs.go

```diff
@@ -0,0 +1,392 @@
+		return errors.New("ValidateRPC must be set")
+	}
+	if e.OnIncomingRPC == nil {
+		return errors.New("OnIncomingRPC must be set")
+	}
+
+	return nil
+}
+
+func (e *PartialMessageExtension) PublishPartial(topic string, partial PartialMessage, opts PublishOptions) error {
+	iwant, err := partial.MissingParts()
+	if err != nil {
+		return err
+	}
+	ihave, err := partial.AvailableParts()
+	if err != nil {
+		return err
+	}
+	groupID := partial.GroupID()
+
+	state, err := e.groupState(topic, groupID)
+	if err != nil {
+		return err
+	}
+
+	state.groupTTL = max(e.GroupTTLByHeatbeat, minGroupTTL)
+
+	var eagerPush *pb.PartialMessage
+	if len(opts.EagerPush) > 0 {
+		eagerPush = &pb.PartialMessage{
+			Data: opts.EagerPush,
+		}
+	}
+
+	var partialIHAVE *pb.PartialIHAVE
+	if len(ihave) > 0 && !bytes.Equal(ihave, state.lastHave) {
+		state.lastHave = ihave
+		partialIHAVE = &pb.PartialIHAVE{
+			Metadata: ihave,
+		}
+	}
+
+	var partialIWANT *pb.PartialIWANT
+	if len(iwant) > 0 {
+		partialIWANT = &pb.PartialIWANT{
+			Metadata: iwant,
+		}
+	}
+
+	peers := opts.PublishToPeers
+	if peers == nil {
+		peers = slices.Collect(e.router.MeshPeers(topic))
+	}
+	e.Logger.Debug("Sending partial messages to peers", "peers", peers, "topic", topic, "group", groupID)
+	for _, p := range peers {
+		log := e.Logger.With("peer", p)
+
+		var rpc pb.PartialMessagesExtension
+		var sendRPC bool
+		var inResponseToIWant bool
+
+		peerState, peerStateOk := state.peerState[p]
+		if !peerStateOk {
+			peerState = &partialMessagePeerState{}
+			state.peerState[p] = peerState
+		}
+		// Try to fulfill any outstanding PartialIWants
+		if peerStateOk && len(peerState.wants) > 0 {
+			// This peer has previously asked for a certain part. We'll give
+			// them what we can.
+			pm, rest, err := partial.PartialMessageBytesFromMetadata(peerState.wants)
+			if err != nil {
+				if e.Logger != nil {
+					e.Logger.Warn("partial message extension failed to get partial message bytes", "error", err)
+				}
+				// Possibly a bad request, we'll delete the request as we will likely error next time we try to handle it
+				state.clearPeerWants(p)
+				continue
+			}
+			if len(rest) == 0 {
+				// We've sent all the parts they asked for
+				state.clearPeerWants(p)
+			} else {
+				// Track what we haven't sent them yet
+				peerState.wants = rest
+			}
+			if len(pm) == 0 {
+				// We don't have anything useful for them
+				continue
+			}
+
+			log.Debug("Respond to peer's IWant")
+			sendRPC = true
+			rpc.Message = &pb.PartialMessage{
+				Data: pm,
+			}
+			inResponseToIWant = true
+		}
+
+		if !inResponseToIWant && eagerPush != nil && !bytes.Equal(ihave, peerState.has) {
+			sendRPC = true
+			log.Debug("Eager pushing")
+			rpc.Message = eagerPush
+		}
+
+		// Only send IHAVE if they don't have what we have and they haven't sent
+		// us an IDONTWANT
+		if !inResponseToIWant && partialIHAVE != nil {
+			if !peerStateOk || !bytes.Equal(ihave, peerState.sentHas) {
```

Comment:
As commented here
```
		// Only send IHAVE if they don't have what we have and they haven't sent
		// us an IDONTWANT
```
I think this uses a wrong field. Can you confirm?
```suggestion
			if !peerStateOk || !bytes.Equal(ihave, peerState.has) {
```

Reply:
The comment was wrong. Updated.